phylo2vec/src/tree_vec/ops/avl.rs

```diff
@@ -43,7 +44,7 @@ impl AVLTree {
         }
     }
 
-    fn update(n: &mut Box<Node>) {
+    fn update_height_and_size(n: &mut Node) {
```

Comment:
Hmm... is this the right type for this function as I think you get `Box<Node>` when you `x.right.as_mut().unwrap()`. Also, `Box` smart pointer stores the data on the Heap rather than the stack: https://doc.rust-lang.org/book/ch15-01-box.html. Also, since this function is recursive, it might be better to use `Box`. I think there might be copying happening if you don't keep this as Box since now data would have to be copied over from heap to the stack.

Reply:
Okay! The Rust analyzer had flagged this as an anti-pattern so I'd changed the type per its [suggestion](https://app.deepsource.com/directory/analyzers/rust/issues/RS-W1050). `Any function that accepts &T as an argument, will also accept &Box<T>. Box<T> implements Deref<Target = T>, which allows any &Box<T> to automatically dereference to a &T.`

Reply:
Correct me if I'm wrong in my interpretation of this [source code](https://doc.rust-lang.org/src/alloc/boxed.rs.html#2373) - but it looks like rust's special dereference implementation for Box<T> doesn't copy to the stack?

Reply:
There's no de-referencing happening within this function. After some look into this, I suppose it doesn't really matter in this case to use `&Box<Node>` or `&Node`. It looks like since they're both references to a known value stored in either stack or heap, it's totally fine to do so (referencing `Box<Node>` as `&Node`. Box doesn't really hold any special thing other than the fact that the underlying data is stored in heap. At the end of the day, since we are just passing around references, no performance is being hit as there are no de-referencing happening.